README
otrtool aims to provide an open source tool to deal with otrkey-files from
onlinetvrecorder.com. It can decrypt them using keys it fetches from OTR's
servers while authenticating with user-supplied login credentials.

== DEPENDENCIES ==

* libmcrypt             for blowfish
* libcurl / curl        for the server request

== BUILDING ==

* execute `make` (or `gmake` where GNU Make is not the default)

* Copy the following files to the places they belong:
    ./otrtool           the main binary
    ./otrtool.1.gz      the gzip'ed manpage
    ./LICENSE           the (rights waiver) license

* `make install` will install the binary and the manpage under /usr/local.
  Use `make install PREFIX=...` to install elsewhere.

* If you didn't install libmcrypt and libcurl in the usual places,
  specify where the headers and libraries are:
  `CFLAGS="-I/path/to/include" LDFLAGS="-L/path/to/lib" make`

== PORTABILITY ==

* otrtool is developed and tested primarily under GNU/Linux.
  If it turns out not to be portable to your unix-like operating system
  of choice, please file a bug report.

* The Makefile is not portable and relies on GNU Make features.

== LINKS ==

project homepage:       http://otrtool.github.io/otrtool/
github repository:      http://github.com/otrtool/otrtool
bug tracker:            http://github.com/otrtool/otrtool/issues

lem's tk/perl GUI:      http://github.com/Lem/otrtool-gui

== CONTACT ==

E-Mail:                 eshrdlu/AT/yandex/DOT/com (de, en)

== REPORTING BUGS ==

For bug reports please do NOT run otrtool with the "-v"-option because it
outputs personal data like your email/password, and/or hashes of them.

== VERSION NUMBERS ==

The version numbers follow the scheme major.minor.bugfixes.
This means that as long as /bugfixes/ increases the software gets more stable.
/Major/ is only incremented on great architectural changes.

== HISTORY ==

In April, 2010, eddy14 reverse-engineered the otrkey-file-format and OTRs
Client-Server-protocol. The blog post describing his work (German):
original URL (dead link):
    http://41yd.de/blog/2010/04/18/otrkey-breaker/
mirror:
    http://pyropeter.eu/41yd.de/blog/2010/04/18/otrkey-breaker/
He also wrote a proof-of-concept tool to decrypt OTRKEYs. PyroPeter rewrote
it in C and without licensing issues.

== LICENSE ==

otrtool uses the CC0 license.

For the license text see the LICENSE file or
http://creativecommons.org/publicdomain/zero/1.0/legalcode


CHANGELOG
For detailed information, see the Git log.

1.3.1 [2022-12-24] (only in stable branch)
* cherry-picked multi-threading bug fix

1.3.0 [2021-01-30]
* multi-threaded decryption

1.2.1 [2017-09-21]
* server IP address updated

1.2.0 [2016-03-01]
* Multiple input files
* Option to unlink after decryption
* 'Verify only' mode of operation

1.1.2 [2017-09-21] (only in stable branch)
* cherry-picked fix from 1.2.1

1.1.1 [2015-02-14] (only in stable branch)
* fixes an issue regarding server communication

1.1.0 [2013-07-15]
* Integrity check before and after decryption
* Improved decryption performance
* Passphrase cache
* Output to a pipe is now supported

1.0.2 [2012-07-12]
* Large file support
* fixes an issue regarding binary constants

1.0.1 [2011-07-12]
* minor changes

1.0 [2010-11-05]
* Offers to overwrite destination file
* Piping in otrkeys using tail -f now works
* Nicer error messages ;-)

0.9 [2010-10-17]
* otrtool -h slimmed down to "important" commands



src/md5.h
/*
 * This is an OpenSSL-compatible implementation of the RSA Data Security, Inc.
 * MD5 Message-Digest Algorithm (RFC 1321).
 *
 * Homepage:
 * http://openwall.info/wiki/people/solar/software/public-domain-source-code/md5
 *
 * Author:
 * Alexander Peslyak, better known as Solar Designer <solar at openwall.com>
 *
 * This software was written by Alexander Peslyak in 2001.  No copyright is
 * claimed, and the software is hereby placed in the public domain.
 * In case this attempt to disclaim copyright and place the software in the
 * public domain is deemed null and void, then the software is
 * Copyright (c) 2001 Alexander Peslyak and it is hereby released to the
 * general public under the following terms:
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted.
 *
 * There's ABSOLUTELY NO WARRANTY, express or implied.
 *
 * See md5.c for more information.
 */

#ifdef HAVE_OPENSSL
#include <openssl/md5.h>
#elif !defined(_MD5_H)
#define _MD5_H

/* Any 32-bit or wider unsigned integer data type will do */
typedef unsigned int MD5_u32plus;

typedef struct {
	MD5_u32plus lo, hi;
	MD5_u32plus a, b, c, d;
	unsigned char buffer[64];
	MD5_u32plus block[16];
} MD5_CTX;

extern void MD5_Init(MD5_CTX *ctx);
extern void MD5_Update(MD5_CTX *ctx, void *data, unsigned long size);
extern void MD5_Final(unsigned char *result, MD5_CTX *ctx);
extern unsigned char *MD5(const unsigned char *d, unsigned long n,
													unsigned char *md);

#endif


src/blowfish.c
/*
  Blowfish algorithm, as described in [1],
  implemented in C for the otrtool project.

  ** Modified endianess **
  Standard Blowfish is big-endian, left-to-right
  (i.e. left = first = most significant).
  otrtool uses a modified algorithm, in which 32-bit words of the data and IV
  are interpreted as little-endian (everything else stays the same).

  [1]
  https://www.schneier.com/academic/archives/1994/09/description_of_a_new.html
*/

#include <string.h>
#include "blowfish.h"
#include "blowfish_const.h"
#define SWAP(x,y,tmp) do { (tmp) = (x); (x) = (y); (y) = (tmp); } while(0)

static uint32_t F(const bf_state_t *st, uint32_t xL)
{
  uint8_t a, b, c, d;

  a = (xL >> 24) & 0xFF;
  b = (xL >> 16) & 0xFF;
  c = (xL >> 8) & 0xFF;
  d = xL & 0xFF;

  return ((st->S[0][a] + st->S[1][b]) ^ st->S[2][c]) + st->S[3][d];
}

#define BF_ENCRYPT(st, xL, xR) do { \
  uint32_t tmp;                     \
  int i;                            \
                                    \
  for (i=0; i<16; i++) {            \
    xL = xL ^ st->P[i];             \
    xR = F(st, xL) ^ xR;            \
    SWAP(xL, xR, tmp);              \
  }                                 \
  SWAP(xL, xR, tmp);                \
  xR = xR ^ st->P[16];              \
  xL = xL ^ st->P[17];              \
} while(0)

#define BF_DECRYPT(st, xL, xR) do { \
  uint32_t tmp;                     \
  int i;                            \
                                    \
  for (i=17; i>1; i--) {            \
    xL = xL ^ st->P[i];             \
    xR = F(st, xL) ^ xR;            \
    SWAP(xL, xR, tmp);              \
  }                                 \
  SWAP(xL, xR, tmp);                \
  xR = xR ^ st->P[1];               \
  xL = xL ^ st->P[0];               \
} while(0)

void bf_init(bf_state_t *st, void *key_, size_t keylen)
{
  uint8_t *key = key_;
  int i, j;
  size_t ki;
  uint32_t k;
  uint32_t xL, xR;

  if (keylen < 1 || keylen > 56)
    return;

  /* step 1 */
  memcpy(st->S, S0, sizeof(S0));
  memcpy(st->P, P0, sizeof(P0));

  /* step 2 */
  ki = 0;
  for (i=0; i<18; i++) {
    k = 0;
    for (j=0; j<4; j++) {
      k = (k << 8) | key[ki];
      if (++ki >= keylen) ki = 0;
    }
    st->P[i] ^= k;
  }

  /* steps 3 to 7 */
  xL = xR = 0;
  for (i=0; i<18; i+=2) {
    BF_ENCRYPT(st, xL, xR);
    st->P[i] = xL;
    st->P[i+1] = xR;
  }
  for (i=0; i<4; i++) {
    for (j=0; j<256; j+=2) {
      BF_ENCRYPT(st, xL, xR);
      st->S[i][j] = xL;
      st->S[i][j+1] = xR;
    }
  }
}

/* Helper functions for endianness conversion
   They are hopefully optimized out on little-endian platforms. */
static uint16_t load_uint16le(uint8_t *b) {
  return (uint16_t)b[1] << 8 | b[0];
}
static uint32_t load_uint32le(uint8_t *b) {
  return (uint32_t)load_uint16le(b+2) << 16 | load_uint16le(b);
}
static void store_uint16le(uint8_t *b, uint16_t x) {
  b[1] = x >> 8; b[0] = x;
}
static void store_uint32le(uint8_t *b, uint32_t x) {
  store_uint16le(b+2, x >> 16); store_uint16le(b, x);
}

/*
  High-level functions for encryption/decryption in ECB and CBC mode
  No padding of input data is performed.
*/

void bf_ecb_decrypt(const bf_state_t *st, void *data_, size_t len)
{
  uint8_t *data = data_;
  uint32_t xL, xR;
  size_t i;

  for (i=0; i<=len-8; i+=8) {
    xL = load_uint32le(data+i);
    xR = load_uint32le(data+i+4);
    BF_DECRYPT(st, xL, xR);
    store_uint32le(data+i, xL);
    store_uint32le(data+i+4, xR);
  }
}

void bf_cbc_encrypt(const bf_state_t *st, void *data_, size_t len, void *iv_)
{
  uint8_t *data = data_;
  uint8_t *iv = iv_;
  uint32_t xL, xR;
  uint32_t ivL, ivR;
  size_t i;

  ivL = load_uint32le(iv);
  ivR = load_uint32le(iv+4);

  for (i=0; i<=len-8; i+=8) {
    xL = load_uint32le(data+i) ^ ivL;
    xR = load_uint32le(data+i+4) ^ ivR;
    BF_ENCRYPT(st, xL, xR);
    store_uint32le(data+i, xL);
    store_uint32le(data+i+4, xR);
    ivL = xL;
    ivR = xR;
  }

  store_uint32le(iv, ivL);
  store_uint32le(iv+4, ivR);
}

void bf_cbc_decrypt(const bf_state_t *st, void *data_, size_t len, void *iv_)
{
  uint8_t *data = data_;
  uint8_t *iv = iv_;
  uint32_t xL, xR;
  uint32_t ivL, ivR, ivLn, ivRn;
  size_t i;

  ivL = load_uint32le(iv);
  ivR = load_uint32le(iv+4);

  for (i=0; i<=len-8; i+=8) {
    xL = ivLn = load_uint32le(data+i);
    xR = ivRn = load_uint32le(data+i+4);
    BF_DECRYPT(st, xL, xR);
    store_uint32le(data+i, xL ^ ivL);
    store_uint32le(data+i+4, xR ^ ivR);
    ivL = ivLn;
    ivR = ivRn;
  }

  store_uint32le(iv, ivL);
  store_uint32le(iv+4, ivR);
}


src/blowfish_gen.c
/*
  Blowfish for otrtool

  Generate the 8336 hexadecimal digits of pi that are
  needed to initialize the Blowfish algorithm
  using the Bailey-Borwein-Plouffe formula
  as described in: https://giordano.github.io/blog/2017-11-21-hexadecimal-pi/

  This little helper program is only used at compile-time.
*/
#include <stdio.h>
#include <stdint.h>

uint32_t modpow(uint32_t b, uint32_t e, uint32_t m)
{
  uint32_t x = 1;

  while (e > 0) {
    if (e % 2 == 0) {
      b = (uint64_t)b*b % m;
      e /= 2;
    }
    else {
      x = (uint64_t)b * x % m;
      e -= 1;
    }
  }
  return x;
}

double s(int n, int j)
{
  int k;
  double x = 0.0;

  for (k=0; k <= n; k++) {
    x += (double)modpow(16, n-k, 8*k+j) / (8*k+j);
    if (x > 1.0) x -= 1.0;
  }

  double d = 1.0;
  double x0 = -1.0;
  for (; x - x0 > 1e-12; k++) {
    x0 = x;
    d /= 16;
    x += d / (8.0*k+j);
  }

  return x;
}

#define TICKS_PER_PERCENT ((8336ul+1)*8336/2)
unsigned long ticks = 0;
int percent = 0;

int pi_digit16(int n)
{
  double d = 16*(4*s(n,1) - 2*s(n,4) - s(n,5) - s(n,6));
  if (d < 0) d -= 1.0;

  ticks += 100ul*(n+1);
  if (ticks >= TICKS_PER_PERCENT) {
    ticks -= TICKS_PER_PERCENT;
    fprintf(stderr, "%d%%\r", ++percent);
    fflush(stderr);
  }

  return (unsigned)(int)d % 16;
}

void print_constants(int *digi, int n, int indent)
{
  int i, j;
  uint32_t x;

  for (i=0; i<n; i++) {
    if (i % 4 == 0) {
      if (i > 0)
        printf(",\n");
      printf("%*c", indent, ' ');
    }
    else
      printf(", ");
    for (x=0, j=0; j<8; j++)
      x = 16*x + pi_digit16((*digi)++);
    printf("0x%.8Xul", x);
  }
}

int main(void)
{
  int n = 0;
  int i;

  printf(
    "/* initial subkeys and S-boxes for Blowfish\n"
    "   generated by blowfish_gen.c */\n"
    "static const uint32_t P0[18] = {\n");
  print_constants(&n, 18, 2);
  printf(
    "\n};\n"
    "static const uint32_t S0[4][256] = {\n");
  for (i=0; i<4; i++) {
    printf("  {\n");
    print_constants(&n, 256, 4);
    printf("\n  }%s\n", (i < 3) ? "," : "");
  }
  printf("};\n");
}


src/md5.c
/*
 * This is an OpenSSL-compatible implementation of the RSA Data Security, Inc.
 * MD5 Message-Digest Algorithm (RFC 1321).
 *
 * Homepage:
 * http://openwall.info/wiki/people/solar/software/public-domain-source-code/md5
 *
 * Author:
 * Alexander Peslyak, better known as Solar Designer <solar at openwall.com>
 *
 * This software was written by Alexander Peslyak in 2001.  No copyright is
 * claimed, and the software is hereby placed in the public domain.
 * In case this attempt to disclaim copyright and place the software in the
 * public domain is deemed null and void, then the software is
 * Copyright (c) 2001 Alexander Peslyak and it is hereby released to the
 * general public under the following terms:
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted.
 *
 * There's ABSOLUTELY NO WARRANTY, express or implied.
 *
 * (This is a heavily cut-down "BSD license".)
 *
 * This differs from Colin Plumb's older public domain implementation in that
 * no exactly 32-bit integer data type is required (any 32-bit or wider
 * unsigned integer data type will do), there's no compile-time endianness
 * configuration, and the function prototypes match OpenSSL's.  No code from
 * Colin Plumb's implementation has been reused; this comment merely compares
 * the properties of the two independent implementations.
 *
 * The primary goals of this implementation are portability and ease of use.
 * It is meant to be fast, but not as fast as possible.  Some known
 * optimizations are not included to reduce source code size and avoid
 * compile-time configuration.
 */

#ifndef HAVE_OPENSSL

#include <stdlib.h>
#include <string.h>

#include "md5.h"

/*
 * The basic MD5 functions.
 *
 * F and G are optimized compared to their RFC 1321 definitions for
 * architectures that lack an AND-NOT instruction, just like in Colin Plumb's
 * implementation.
 */
#define F(x, y, z)			((z) ^ ((x) & ((y) ^ (z))))
#define G(x, y, z)			((y) ^ ((z) & ((x) ^ (y))))
#define H(x, y, z)			((x) ^ (y) ^ (z))
#define I(x, y, z)			((y) ^ ((x) | ~(z)))

/*
 * The MD5 transformation for all four rounds.
 */
#define STEP(f, a, b, c, d, x, t, s) \
	(a) += f((b), (c), (d)) + (x) + (t); \
	(a) = (((a) << (s)) | (((a) & 0xffffffff) >> (32 - (s)))); \
	(a) += (b);

/*
 * SET reads 4 input bytes in little-endian byte order and stores them
 * in a properly aligned word in host byte order.
 *
 * The check for little-endian architectures that tolerate unaligned
 * memory accesses is just an optimization.  Nothing will break if it
 * doesn't work.
 */
#if defined(__i386__) || defined(__x86_64__) || defined(__vax__)
#define SET(n) \
	(*(MD5_u32plus *)&ptr[(n) * 4])
#define GET(n) \
	SET(n)
#else
#define SET(n) \
	(ctx->block[(n)] = \
	(MD5_u32plus)ptr[(n) * 4] | \
	((MD5_u32plus)ptr[(n) * 4 + 1] << 8) | \
	((MD5_u32plus)ptr[(n) * 4 + 2] << 16) | \
	((MD5_u32plus)ptr[(n) * 4 + 3] << 24))
#define GET(n) \
	(ctx->block[(n)])
#endif

/*
 * This processes one or more 64-byte data blocks, but does NOT update
 * the bit counters.  There are no alignment requirements.
 */
static void *body(MD5_CTX *ctx, void *data, unsigned long size)
{
	unsigned char *ptr;
	MD5_u32plus a, b, c, d;
	MD5_u32plus saved_a, saved_b, saved_c, saved_d;

	ptr = data;

	a = ctx->a;
	b = ctx->b;
	c = ctx->c;
	d = ctx->d;

	do {
		saved_a = a;
		saved_b = b;
		saved_c = c;
		saved_d = d;

/* Round 1 */
		STEP(F, a, b, c, d, SET(0), 0xd76aa478, 7)
		STEP(F, d, a, b, c, SET(1), 0xe8c7b756, 12)
		STEP(F, c, d, a, b, SET(2), 0x242070db, 17)
		STEP(F, b, c, d, a, SET(3), 0xc1bdceee, 22)
		STEP(F, a, b, c, d, SET(4), 0xf57c0faf, 7)
		STEP(F, d, a, b, c, SET(5), 0x4787c62a, 12)
		STEP(F, c, d, a, b, SET(6), 0xa8304613, 17)
		STEP(F, b, c, d, a, SET(7), 0xfd469501, 22)
		STEP(F, a, b, c, d, SET(8), 0x698098d8, 7)
		STEP(F, d, a, b, c, SET(9), 0x8b44f7af, 12)
		STEP(F, c, d, a, b, SET(10), 0xffff5bb1, 17)
		STEP(F, b, c, d, a, SET(11), 0x895cd7be, 22)
		STEP(F, a, b, c, d, SET(12), 0x6b901122, 7)
		STEP(F, d, a, b, c, SET(13), 0xfd987193, 12)
		STEP(F, c, d, a, b, SET(14), 0xa679438e, 17)
		STEP(F, b, c, d, a, SET(15), 0x49b40821, 22)

/* Round 2 */
		STEP(G, a, b, c, d, GET(1), 0xf61e2562, 5)
		STEP(G, d, a, b, c, GET(6), 0xc040b340, 9)
		STEP(G, c, d, a, b, GET(11), 0x265e5a51, 14)
		STEP(G, b, c, d, a, GET(0), 0xe9b6c7aa, 20)
		STEP(G, a, b, c, d, GET(5), 0xd62f105d, 5)
		STEP(G, d, a, b, c, GET(10), 0x02441453, 9)
		STEP(G, c, d, a, b, GET(15), 0xd8a1e681, 14)
		STEP(G, b, c, d, a, GET(4), 0xe7d3fbc8, 20)
		STEP(G, a, b, c, d, GET(9), 0x21e1cde6, 5)
		STEP(G, d, a, b, c, GET(14), 0xc33707d6, 9)
		STEP(G, c, d, a, b, GET(3), 0xf4d50d87, 14)
		STEP(G, b, c, d, a, GET(8), 0x455a14ed, 20)
		STEP(G, a, b, c, d, GET(13), 0xa9e3e905, 5)
		STEP(G, d, a, b, c, GET(2), 0xfcefa3f8, 9)
		STEP(G, c, d, a, b, GET(7), 0x676f02d9, 14)
		STEP(G, b, c, d, a, GET(12), 0x8d2a4c8a, 20)

/* Round 3 */
		STEP(H, a, b, c, d, GET(5), 0xfffa3942, 4)
		STEP(H, d, a, b, c, GET(8), 0x8771f681, 11)
		STEP(H, c, d, a, b, GET(11), 0x6d9d6122, 16)
		STEP(H, b, c, d, a, GET(14), 0xfde5380c, 23)
		STEP(H, a, b, c, d, GET(1), 0xa4beea44, 4)
		STEP(H, d, a, b, c, GET(4), 0x4bdecfa9, 11)
		STEP(H, c, d, a, b, GET(7), 0xf6bb4b60, 16)
		STEP(H, b, c, d, a, GET(10), 0xbebfbc70, 23)
		STEP(H, a, b, c, d, GET(13), 0x289b7ec6, 4)
		STEP(H, d, a, b, c, GET(0), 0xeaa127fa, 11)
		STEP(H, c, d, a, b, GET(3), 0xd4ef3085, 16)
		STEP(H, b, c, d, a, GET(6), 0x04881d05, 23)
		STEP(H, a, b, c, d, GET(9), 0xd9d4d039, 4)
		STEP(H, d, a, b, c, GET(12), 0xe6db99e5, 11)
		STEP(H, c, d, a, b, GET(15), 0x1fa27cf8, 16)
		STEP(H, b, c, d, a, GET(2), 0xc4ac5665, 23)

/* Round 4 */
		STEP(I, a, b, c, d, GET(0), 0xf4292244, 6)
		STEP(I, d, a, b, c, GET(7), 0x432aff97, 10)
		STEP(I, c, d, a, b, GET(14), 0xab9423a7, 15)
		STEP(I, b, c, d, a, GET(5), 0xfc93a039, 21)
		STEP(I, a, b, c, d, GET(12), 0x655b59c3, 6)
		STEP(I, d, a, b, c, GET(3), 0x8f0ccc92, 10)
		STEP(I, c, d, a, b, GET(10), 0xffeff47d, 15)
		STEP(I, b, c, d, a, GET(1), 0x85845dd1, 21)
		STEP(I, a, b, c, d, GET(8), 0x6fa87e4f, 6)
		STEP(I, d, a, b, c, GET(15), 0xfe2ce6e0, 10)
		STEP(I, c, d, a, b, GET(6), 0xa3014314, 15)
		STEP(I, b, c, d, a, GET(13), 0x4e0811a1, 21)
		STEP(I, a, b, c, d, GET(4), 0xf7537e82, 6)
		STEP(I, d, a, b, c, GET(11), 0xbd3af235, 10)
		STEP(I, c, d, a, b, GET(2), 0x2ad7d2bb, 15)
		STEP(I, b, c, d, a, GET(9), 0xeb86d391, 21)

		a += saved_a;
		b += saved_b;
		c += saved_c;
		d += saved_d;

		ptr += 64;
	} while (size -= 64);

	ctx->a = a;
	ctx->b = b;
	ctx->c = c;
	ctx->d = d;

	return ptr;
}

void MD5_Init(MD5_CTX *ctx)
{
	ctx->a = 0x67452301;
	ctx->b = 0xefcdab89;
	ctx->c = 0x98badcfe;
	ctx->d = 0x10325476;

	ctx->lo = 0;
	ctx->hi = 0;
}

void MD5_Update(MD5_CTX *ctx, void *data, unsigned long size)
{
	MD5_u32plus saved_lo;
	unsigned long used, free;

	saved_lo = ctx->lo;
	if ((ctx->lo = (saved_lo + size) & 0x1fffffff) < saved_lo)
		ctx->hi++;
	ctx->hi += size >> 29;

	used = saved_lo & 0x3f;

	if (used) {
		free = 64 - used;

		if (size < free) {
			memcpy(&ctx->buffer[used], data, size);
			return;
		}

		memcpy(&ctx->buffer[used], data, free);
		data = (unsigned char *)data + free;
		size -= free;
		body(ctx, ctx->buffer, 64);
	}

	if (size >= 64) {
		data = body(ctx, data, size & ~(unsigned long)0x3f);
		size &= 0x3f;
	}

	memcpy(ctx->buffer, data, size);
}

void MD5_Final(unsigned char *result, MD5_CTX *ctx)
{
	unsigned long used, free;

	used = ctx->lo & 0x3f;

	ctx->buffer[used++] = 0x80;

	free = 64 - used;

	if (free < 8) {
		memset(&ctx->buffer[used], 0, free);
		body(ctx, ctx->buffer, 64);
		used = 0;
		free = 64;
	}

	memset(&ctx->buffer[used], 0, free - 8);

	ctx->lo <<= 3;
	ctx->buffer[56] = ctx->lo;
	ctx->buffer[57] = ctx->lo >> 8;
	ctx->buffer[58] = ctx->lo >> 16;
	ctx->buffer[59] = ctx->lo >> 24;
	ctx->buffer[60] = ctx->hi;
	ctx->buffer[61] = ctx->hi >> 8;
	ctx->buffer[62] = ctx->hi >> 16;
	ctx->buffer[63] = ctx->hi >> 24;

	body(ctx, ctx->buffer, 64);

	result[0] = ctx->a;
	result[1] = ctx->a >> 8;
	result[2] = ctx->a >> 16;
	result[3] = ctx->a >> 24;
	result[4] = ctx->b;
	result[5] = ctx->b >> 8;
	result[6] = ctx->b >> 16;
	result[7] = ctx->b >> 24;
	result[8] = ctx->c;
	result[9] = ctx->c >> 8;
	result[10] = ctx->c >> 16;
	result[11] = ctx->c >> 24;
	result[12] = ctx->d;
	result[13] = ctx->d >> 8;
	result[14] = ctx->d >> 16;
	result[15] = ctx->d >> 24;

	memset(ctx, 0, sizeof(*ctx));
}

unsigned char staticmd[16];

unsigned char * MD5(const unsigned char *d, unsigned long n,
										unsigned char *md) {
	/* This function was added by PyroPeter to allow reusal of old
	 * openssl-dependent code */
	
	MD5_CTX ctx;
	memset(&ctx, 0, sizeof(MD5_CTX));
	MD5_Init(&ctx);
	
	MD5_Update(&ctx, (void *) d, n);
	
	unsigned char *result = (md ? md : (unsigned char *) staticmd);
	MD5_Final(result, &ctx);
	return staticmd;
}

#endif
